homeassistant/components/mpchc/media_player.py

```diff
@@ -55,46 +67,78 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     name = config.get(CONF_NAME)
     host = config.get(CONF_HOST)
     port = config.get(CONF_PORT)
+    log = config.get(CONF_LOG)
+    power_sensor = config.get(CONF_POWER_SENSOR)
 
     url = f"{host}:{port}"
 
-    add_entities([MpcHcDevice(name, url)], True)
+    add_entities([MpcHcDevice(hass, name, url, log, power_sensor)], True)
 
 
 class MpcHcDevice(MediaPlayerEntity):
     """Representation of a MPC-HC server."""
 
-    def __init__(self, name, url):
+    def __init__(self, hass, name, url, log, power_sensor):
         """Initialize the MPC-HC device."""
+        self.hass = hass
         self._name = name
         self._url = url
         self._player_variables = {}
         self._available = False
+        self._log = log
+        self._power_sensor_available = False
+        self._power_sensor = power_sensor
+
+    async def async_added_to_hass(self):
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+
+        if self._power_sensor:
+            async_track_state_change(self.hass, self._power_sensor,
+                                     self._async_power_sensor_changed)
+
+    async def _async_power_sensor_changed(self, entity_id, old_state, new_state):
+        """Handle power sensor changes."""
+        if new_state is None:
+            return
+
+        if new_state.state == STATE_ON and self._power_sensor_available == False:
+            self._power_sensor_available = True
+        if new_state.state == STATE_OFF:
+            if self._power_sensor_available == True:
+                self._power_sensor_available = False
+            if self._available == True:
+                self._available = False
 
     def update(self):
         """Get the latest details."""
-        try:
-            response = requests.get(f"{self._url}/variables.html", data=None, timeout=3)
-
-            mpchc_variables = re.findall(r'<p id="(.+?)">(.+?)</p>', response.text)
-
-            for var in mpchc_variables:
-                self._player_variables[var[0]] = var[1].lower()
-            self._available = True
-        except requests.exceptions.RequestException:
-            _LOGGER.error("Could not connect to MPC-HC at: %s", self._url)
-            self._player_variables = {}
-            self._available = False
+        if not self._power_sensor or self._power_sensor and self._power_sensor_available:
+            try:
+                response = requests.get(f"{self._url}/variables.html", data=None, timeout=3)
```

Comment:
The device specific interface code and data parsing needs to be extracted to a standalone library published on PyPI, before we can accept changes to this integration.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices